src/deploy/functions/validate.ts

```diff
@@ -0,0 +1,87 @@
+import * as FirebaseError from "../../error";
+import * as _ from "lodash";
+import * as path from "path";
+import * as clc from "cli-color";
+import * as logger from "../../logger";
+import * as projectPath from "../../projectPath";
+import * as fsutils from "../../fsutils";
+
+// have to require this because no @types/cjson available
+// tslint:disable-next-line
+const cjson = require("cjson");
+
+/**
+ * Check that functions directory exists.
+ * @param cwd Working directory.
+ * @param sourceDirName Relative path to source directory.
+ * @throws { FirebaseError } Functions directory must exist.
+ */
+export function functionsDirectoryExists(cwd: string, sourceDirName: string): void {
+  if (!fsutils.dirExistsSync(projectPath.resolveProjectPath(cwd, sourceDirName))) {
+    const msg =
+      `could not deploy functions because the ${clc.bold('"' + sourceDirName + '"')} ` +
+      `directory was not found. Please create it or specify a different source directory in firebase.json`;
+    throw new FirebaseError(msg);
+  }
+}
+
+/**
+ * Validate function names only contain lower case letters, numbers, and dashes.
```

Comment:
function names can include uppercase right? and underscore

Reply:
Yeah I think so - I've been having trouble finding the docs that define what is a valid function name.

Reply:
Yep, you're right. Changing to "Validate function names only contain letters, numbers, underscores, and hyphens"